### PR TITLE
Custom tasks

### PR DESCRIPTION
**Pull request by bomoko**
This PR introduces a custom task registration system.

# Checklist

- [ ] Affected Issues have been mentioned in the Closing issues section
- [x] Documentation has been written/updated
- [ ] PR title is ready for changelog and subsystem label(s) applied

This PR introduces the ability to define custom tasks to be run against environments in Lagoon.
It supports permission based invocation, and task registration against environments, projects, and groups.

First step in addressing  #1826 and #1385

### CODE REVIEW

File: services/api/src/resources/task/task_definition_resolvers.ts [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    systemLevelTask || advancedTaskWithImage || groupName;
+
+  if (systemLevelTask) {
+    //if they pass this, they can do basically anything
+    //In the first release, we're not actually supporting this
+    //TODO: add checks once images are officially supported - for now, throw an error
+    throw Error('Adding Images and System Wide Tasks are not yet supported');
+  } else if (advancedTaskWithImage) {
+    //We're only going to allow administrators to add these for now ...
+    // await hasPermission('advanced_task','create:advanced');
```
Comment (bomoko) [inline services/api/src/resources/task/task_definition_resolvers.ts]: This should be uncommented

**Comment by tobybellwood** [comment]
I'm seeing an error in the api logs:
```
2021-07-29T04:31:46.909392906Z [2021-07-29 04:31:46] [pubsub-[31merror[39m]: unable to send message to exchange '"api.event.task.updated.DLQ.Exchange"' {undefined}
2021-07-29T04:31:46.910970051Z Unhandled rejection Error: Unable to send message
2021-07-29T04:31:46.910980031Z     at /app/node_modules/rabbitmq-pub-sub/dist/main.js:337:39
2021-07-29T04:31:46.910982538Z     at __NR_wrappedThenHandler (/app/node_modules/newrelic/lib/shim/promise-shim.js:407:26)
2021-07-29T04:31:46.910984396Z     at __NR_wrappedThenHandler (/app/node_modules/newrelic/lib/shim/promise-shim.js:407:26)
2021-07-29T04:31:46.910986325Z     at tryCatcher (/app/node_modules/bluebird/js/release/util.js:16:23)
2021-07-29T04:31:46.910988271Z     at Promise._settlePromiseFromHandler (/app/node_modules/bluebird/js/release/promise.js:517:31)
2021-07-29T04:31:46.910990019Z     at Promise._settlePromise (/app/node_modules/bluebird/js/release/promise.js:574:18)
2021-07-29T04:31:46.910991777Z     at Promise._settlePromise0 (/app/node_modules/bluebird/js/release/promise.js:619:10)
2021-07-29T04:31:46.910993494Z     at Promise._settlePromises (/app/node_modules/bluebird/js/release/promise.js:695:18)
2021-07-29T04:31:46.910995233Z     at _drainQueueStep (/app/node_modules/bluebird/js/release/async.js:138:12)
2021-07-29T04:31:46.910996950Z     at _drainQueue (/app/node_modules/bluebird/js/release/async.js:131:9)
2021-07-29T04:31:46.910998701Z     at Async._drainQueues (/app/node_modules/bluebird/js/release/async.js:147:5)
2021-07-29T04:31:46.911000447Z     at Immediate.Async.drainQueues [as _onImmediate] (/app/node_modules/bluebird/js/release/async.js:17:14)
2021-07-29T04:31:46.911002174Z     at processImmediate (node:internal/timers:464:21)
2021-07-29T04:31:46.911003956Z     at process.topLevelDomainCallback (node:domain:147:15)
2021-07-29T04:31:46.911005707Z     at process.callbackTrampoline (node:internal/async_hooks:128:24)
```
Full API logs attached from a kind/retest run only on the `tasks` test.

Also noted no API audit output (yet) - will need to be added.

[logs-api-zsswn.txt](https://github.com/uselagoon/lagoon/files/6898100/logs-api-zsswn.txt)

**Comment by tobybellwood** [comment]
looking at the error a bit, it seems that the API doesn't resolve the task ID for image tasks properly - they come through as id `0` in the logs, whereas the command tasks have the correct ID

The spec for the task is missing the id
```
spec:
  advancedTask:
    JSONPayload: W10=
    runnerImage: hello-world
  environment:
    environmentType: ''
    id: ''
    name: tasks
    openshiftProjectName: ci-tasks-control-k8s-tasks
    project: ''
  key: 'kubernetes:task:advanced'
```

**Comment by tobybellwood** [comment]
exchange/spec error fixed in [`5a29beb` (#2779)](https://github.com/uselagoon/lagoon/pull/2779/commits/5a29beb36438f3e1d9157c9e206627a5b5855d04)

thanks @shreddedbacon :mage:

**Comment by bomoko** [comment]
@tobybellwood - I came here to fix this precise error :P That's what I mentioned last time, thought it smelled a lot like the issue we fixed with the standard tasks.

**Pull request merged**